hazelcast/src/main/java/com/hazelcast/sql/impl/SqlInternalService.java

```diff
@@ -209,4 +210,33 @@ public QueryOperationHandlerImpl getOperationHandler() {
     public QueryClientStateRegistry getClientStateRegistry() {
         return clientStateRegistry;
     }
+
+    private void prepareParameters(Plan plan, List<Object> params) {
+        assert params != null;
+        QueryParameterMetadata parameterMetadata = plan.getParameterMetadata();
+
+        int parameterCount = parameterMetadata.getParameterCount();
+        if (parameterCount != params.size()) {
+            throw QueryException.error(SqlErrorCode.DATA_EXCEPTION,
+                    "Unexpected parameter count: expected " + parameterCount + ", got " + params.size());
+        }
+
+        for (int i = 0; i < params.size(); ++i) {
+            Object value = params.get(i);
+            if (value == null) {
+                continue;
+            }
+
+            Converter valueConverter = Converters.getConverter(value.getClass());
+            Converter typeConverter = parameterMetadata.getParameterType(i).getConverter();
+            try {
+                value = typeConverter.convertToSelf(valueConverter, value);
+            } catch (RuntimeException e) {
+                throw QueryException.error(SqlErrorCode.DATA_EXCEPTION,
```

Comment:
I would propose to add the type to which we tried to convert. E.g.: 
`Failed to convert parameter at position 1 to VARCHAR: ...`

Reply:
fixed in efba7d68347224dede855a5915fcb96c8a9a3482